zone/spell_effects.cpp

```diff
@@ -3994,6 +3995,12 @@ void Mob::DoBuffTic(const Buffs_Struct &buff, int slot, Mob *caster)
 		case SE_Charm: {
 			if (!caster || !PassCharismaCheck(caster, buff.spellid)) {
 				BuffFadeByEffect(SE_Charm);
+				
+				// Remove from hate list of any NPC's hate list and remove all NPCs this hate list
+				if (IsNPC()) {
+					entity_list.RemoveFromHateLists(this);
+					WipeHateList(true);
+				}
```

Comment:
I don't know the spell code at all but would it make more sense to handle this in the `SE_Charm` handler inside `BuffFadeBySlot`?

There's also already some hate list stuff there (https://github.com/EQEmu/Server/blame/1212ccefef7132221425201bc47ff53d83af1468/zone/spell_effects.cpp#L4422) added in https://github.com/EQEmu/Server/commit/af5cfb9bed7c4fdfbef3dcf42bb39b349cfb3888 so not sure if that is newer than the base for this code and something there might conflict